### PR TITLE
[Indexer] Integrate pocketdex into Tiltfile

### PR DESCRIPTION
**Pull request by bryanchriswhite**
## Summary

- Adds the pocketdex to the tilt environment **if it has been cloned to a sibling directory as "pocketdex"**
- Otherwise, if `localnet_config.yaml` has `indexer.enabled` and `indexer.clone_if_not_present` set to `true` Tilt will clone it to a sibling directory
- Otherwise a stub service titled "Indexer Disabled" under the "Pocketdex" label group explains what changes need to be made to the localnet_config.yaml:

    image

```
Running cmd: echo 'Pocketdex indexer disabled. Set `indexer.enabled` to `true` in `localnet_config.yaml` to enable it.'
Pocketdex indexer disabled. Set `indexer.enabled` to `true` in `localnet_config.yaml` to enable it.
```

```
Running cmd: echo 'Pocketdex repo not found. Set `clone_if_not_present` to `true` in `localnet_config.yaml` to clone the repo.'
Pocketdex repo not found. Set `clone_if_not_present` to `true` in `localnet_config.yaml` to clone the repo.
```

## Issue

- #873 

## Type of change

Select one or more from the following:

- [x] New feature, functionality or library
- [ ] Consensus breaking; add the `consensus-breaking` label if so. See #791 for details
- [ ] Bug fix
- [ ] Code health or cleanup
- [ ] Documentation
- [ ] Other (specify)

## Testing

- [ ] **Documentation**: `make docusaurus_start`; only needed if you make doc changes
- [ ] **Unit Tests**: `make go_develop_and_test`
- [ ] **LocalNet E2E Tests**: `make test_e2e`
- [ ] **DevNet E2E Tests**: Add the `devnet-test-e2e` label to the PR.

## Sanity Checklist

- [ ] I have tested my changes using the available tooling
- [ ] I have commented my code
- [ ] I have performed a self-review of my own code; both comments & source code
- [ ] I create and reference any new tickets, if applicable
- [ ] I have left TODOs throughout the codebase, if applicable

### CODE REVIEW

File: Tiltfile [hunk 1]
```diff
@@ -383,3 +386,34 @@ if localnet_config["rest"]["enabled"]:
...
+                                gql_engine_values_path="./localnet/kubernetes/values-pocketdex-gql-engine.yaml")
+
+if localnet_config["indexer"]["enabled"]:
+    # Check if sibling pocketdex repo exists.
+    pocketdex_repo_exists = local("[ -d '../pocketdex' ] && echo 'true' || echo 'false'")
+    print("POCKETDEX_REPO_EXISTS: ", str(pocketdex_repo_exists).strip())
+    if str(pocketdex_repo_exists).strip() == "false":
+        if localnet_config["indexer"]["clone_if_not_present"]:
+            print("Cloning pocketdex repo")
+            local("git clone https://github.com/pokt-network/pocketdex --branch chore/tilt ../pocketdex")
```
Comment (bryanchriswhite) [inline Tiltfile:406]: TODO: Remove `--branch chore/tilt` flag once https://github.com/pokt-network/pocketdex/pull/23 is merged.

```suggestion
            local("git clone https://github.com/pokt-network/pocketdex ../pocketdex")
```

File: Tiltfile [hunk 2]
```diff
@@ -383,3 +386,34 @@ if localnet_config["rest"]["enabled"]:
...
+            load_pocketdex()
+        else:
+            print("NOT CLONING")
+            local_resource("Indexer Disabled",
```
Comment (bryanchriswhite) [inline Tiltfile:410]: Factor this `local_resource` out to a function that takes the string to print as an argument.

File: Tiltfile [hunk 3]
```diff
@@ -383,3 +386,34 @@ if localnet_config["rest"]["enabled"]:
...
+                           "echo 'Pocketdex repo not found. Set `clone_if_not_present` to `true` in `localnet_config.yaml` and restart tilt to clone the repo.'",
```
Comment (bryanchriswhite) [inline Tiltfile:411]:
```suggestion
                           "echo 'Pocketdex repo not found. Set `clone_if_not_present` to `true` in `localnet_config.yaml` to clone the repo.'",
```

Also consider interpolating the expected path.

File: localnet/kubernetes/values-pocketdex-gql-engine.yaml [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+  NODE_ENV: development
+  # Testnet
+  ENDPOINT: "https://testnet-validated-validator-rpc.poktroll.com"
+  # LocalNet
+  # ENDPOINT: "http://validator-poktroll-validator.default.svc.cluster.local:26657"
+  CHAIN_ID: "poktroll"
+  # db schema name
+  DB_SCHEMA: "localnet"
+  # indexer endpoint
+#  indexer_endpoint: "http://pocketdex-indexer.default.svc.cluster.local:3000"
```
Comment (bryanchriswhite) [inline localnet/kubernetes/values-pocketdex-gql-engine.yaml:11]:
```suggestion
  indexer_endpoint: "http://indexer-service:3000"
```

File: Tiltfile [hunk 5]
```diff
@@ -383,3 +386,37 @@ if localnet_config["rest"]["enabled"]:
...
+
+### Pocketdex Shannon Indexer
+def pocketdex_disabled(reason):
+    local_resource("Indexer Disabled",
+                   "echo '{}'".format(reason),
+                   labels=["Pocketdex"])
+
+def load_pocketdex():
+    pocketdex_tilt = load_dynamic("../pocketdex/tiltfiles/pocketdex.tilt")
+    pocketdex_tilt["pocketdex"]("../pocketdex",
```
Comment (bryanchriswhite) [inline Tiltfile:398]:
```suggestion
    pocketdex_tilt["pocketdex"](localnet_config["indexer"]["pocketdex_path"],
```

**Comment by Olshansk** [comment]
@bryanchriswhite What will it take to make this ready for review?

**Comment by bryanchriswhite** [comment]
@Olshansk I want to check for any cleanup that needs to be done in [pocketdex#23](https://github.com/pokt-network/pocketdex/pull/23) and it needs to be merged to main before this can be merged.

File: Tiltfile [hunk 6]
```diff
@@ -383,3 +389,44 @@ if localnet_config["rest"]["enabled"]:
...
+    # If it does, load the pocketdex.tilt file from the sibling repo.
+    # Otherwise check the `indexer.clone_if_not_present` flag in `localnet_config.yaml`
+    # and EITHER clone pocketdex to ../pocketdex OR prints a message if true or false,
+    # respectively.
+    pocketdex_repo_exists = local("[ -d {} ] && echo 'true' || echo 'false'".format(pocketdex_root_path))
+    if str(pocketdex_repo_exists).strip() == "false":
+        if localnet_config["indexer"]["clone_if_not_present"]:
+            print("Cloning pocketdex repo")
+            # TODO_INVESTIGATE: https://github.com/tilt-dev/tilt-extensions/tree/master/git_resource
+            local("git clone https://github.com/pokt-network/pocketdex --branch chore/tilt ../pocketdex")
```
Comment (bryanchriswhite) [inline Tiltfile:424]:
```suggestion
            local("git clone https://github.com/pokt-network/pocketdex --branch main ../pocketdex")
```

File: Tiltfile [hunk 7]
```diff
@@ -383,3 +396,114 @@ if localnet_config["rest"]["enabled"]:
...
+
+# Check if sibling pocketdex repo exists.
+# If it does, load the pocketdex.tilt file from the sibling repo.
+# Otherwise check the `indexer.clone_if_not_present` flag in `localnet_config.yaml` and EITHER:
+#   1. clone pocketdex to ../pocketdex
+#   -- OR --
+#   2. Prints a message if true or false
+if localnet_config["indexer"]["enabled"]:
+    pocketdex_repo_exists = local("[ -d {} ] && echo 'true' || echo 'false'".format(pocketdex_root_path))
+    if pocketdex_repo_exists == "false":
```
Comment (bryanchriswhite) [inline Tiltfile:494]:
```suggestion
    if pocketdex_repo_exists() == "false":
```

**Review by Olshansk — DISMISSED** [review]
@bryanchriswhite Pushed my changes after review + offline discussion in [1].

Requesting changes as a checkpoint for when you need me to take another look at this.

[1] https://discord.com/channels/824324475256438814/1296487759745253408

File: docusaurus/docs/develop/localnet/observability.md [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+
+This document is a living WIP and assumes you are familiar with the LocalNet environment.
+
+:::
+
+## Pocketdex <!-- omit in toc -->
+
+[Pocketdex](https://github.com/pokt-network/pocketdex/), the poktroll indexer starts up as part of the default LocalNet.
+
+### Usage
```
Comment (bryanchriswhite) [inline docusaurus/docs/develop/localnet/observability.md:22]:
```suggestion
### Usage

#### See [pocketdex docs](https://github.com/pokt-network/pocketdex?tab=readme-ov-file#usage--query-docs) for more.
```

File: Tiltfile [hunk 9]
```diff
@@ -55,9 +55,20 @@ localnet_config_defaults = {
     "rest": {
         "enabled": True,
     },
+
+    # NOTE: git submodule usage was explicityly avoided to reduce environment complexity.
```
Comment (bryanchriswhite) [inline Tiltfile:59]:
```suggestion
    # NOTE: git submodule usage was explicitly avoided to reduce environment complexity.
```

File: tiltfiles/pocketdex.tilt [hunk 10]
```diff
@@ -0,0 +1,73 @@
+load("./git.tilt",
```
Comment (Olshansk) [inline tiltfiles/pocketdex.tilt:1]: Thank you for splitting this into multiple sections. I do which Github would let us customize syntax highlighting though...

File: tiltfiles/pocketdex.tilt [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+    gql_engine_values_path = os.path.join(".", "localnet", "kubernetes", "values-pocketdex-gql-engine.yaml")
+    pocketdex_tilt["pocketdex"](pocketdex_root_path,
+                                genesis_file_name="localnet.json",
+                                postgres_values_path=postgres_values_path,
+                                pgadmin_values_path=pgadmin_values_path,
+                                indexer_values_path=indexer_values_path,
+                                gql_engine_values_path=gql_engine_values_path)
+
+
+# Check if sibling pocketdex repo exists.
```
Comment (Olshansk) [inline tiltfiles/pocketdex.tilt:50]:
```suggestion
# check_and_load_pocketdex check if sibling pocketdex repo exists.
```

File: tiltfiles/git.tilt [hunk 12]
```diff
@@ -0,0 +1,53 @@
+# dir_repo_exists returns true if the dir_path exists.
```
Comment (Olshansk) [inline tiltfiles/git.tilt:1]:
```suggestion
# dir_exists returns true if the dir_path exists.
```

File: Tiltfile [hunk 13]
```diff
@@ -383,3 +395,14 @@ if localnet_config["rest"]["enabled"]:
     print("REST enabled: " + str(localnet_config["rest"]["enabled"]))
     deployment_create("rest", image="davarski/go-rest-api-demo")
     k8s_resource("rest", labels=["data_nodes"], port_forwards=["10000"])
+
+### Pocketdex Shannon Indexer
+load("./tiltfiles/pocketdex.tilt", "check_and_load_pocketdex")
```
Comment (Olshansk) [inline Tiltfile:400]: For context, I just reviewed the comments in the other .tilt files (not the code), but lgtm

File: docusaurus/docs/develop/localnet/observability.md [hunk 14]
```diff
@@ -0,0 +1,72 @@
+---
```
Comment (Olshansk) [inline docusaurus/docs/develop/localnet/observability.md:1]: Ty for updating this w/ all the details!

File: docusaurus/docs/develop/localnet/observability.md [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+set schema 'localnet';
+\dt
+select * from accounts limit 10; # Example query
+```
+
+### Debugging
+
+#### Port already in use
+
+If you go to [http://localhost:10350/r/Postgres/overview] and see the following error:
```
Comment (Olshansk) [inline docusaurus/docs/develop/localnet/observability.md:56]:
```suggestion
If you go to [http://localhost:10350/r/Postgres/overview](http://localhost:10350/r/Postgres/overview) and see the following error:
```

**Review by Olshansk — APPROVED** [review]

**Review by okdas — COMMENTED** [review]
You've upped `Tiltfile` game in this repo! :D Now we're going to need to break our main huge `Tiltfile` into smaller ones. This is great, thank you!

File: tiltfiles/git.tilt [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+def dir_exists(dir_path):
+    return str(
+        local("[ -d {} ] && echo 'true' || echo 'false'".format(dir_path),
```
[the remark this answers is not in the document]
Comment (bryanchriswhite) [reply tiltfiles/git.tilt:4]: :eyes: I was not aware of this :sweat_smile: - thanks for enlightening me! :rocket: 

Will simplify! :raised_hands:

**Pull request merged**